core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/EnvelopeWrappedSdkPublisher.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2010-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.async;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiFunction;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.annotations.ThreadSafe;
+import software.amazon.awssdk.core.async.SdkPublisher;
+
+/**
+ * Publisher implementation that wraps the content of another publisher in an envelope with an optional prefix (or
+ * header) and suffix (or footer). The prefix content will be prepended to the first published object from the
+ * wrapped publisher, and the suffix content will be published when the wrapped publisher signals completion.
+ * <p>
+ * The envelope prefix will not be published until the wrapped publisher publishes something or is completed.
+ * The envelope suffix will not be published until the wrapped publisher is completed.
+ * <p>
+ * This class can be used in an asynchronous interceptor in the AWS SDK to wrap content around the incoming
+ * bytestream from a response.
+ * <p>
+ * A function must be supplied that can be used to concatinate the envelope content to the content being published by
+ * the wrapped publisher. Example usage:
+ * {@code
+ *    Publisher<String> wrappedPublisher = ContentEnvelopeWrappingPublisher.of(publisher, "S", "E", (s1, s2) -> s1 + s2);
+ * }
+ * If publisher publishes a single string "1", wrappedPublisher will publish "S1" (prepending the envelop prefix). If
+ * publisher then publishes a second string "2", wrappedPublisher will then publish "2" (no added content). If
+ * publisher then completes, wrappedPublisher will then publish "E" and then complete.
+ *
+ * @param <T> The type of objects being published
+ */
+
+@ThreadSafe
+@SdkInternalApi
+public class EnvelopeWrappedSdkPublisher<T> implements SdkPublisher<T> {
```

Comment:
We probably should add reactive-stream tcl test for it as well.

https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md#reactive-streams

Reply:
As discussed offline, this implementation will never be truly compliant as crudely acts as a man-in-the-middle between an external publisher and subscriber. In this case we control the publisher side, so we're able to hide/mitigate the way we are inserting data where there should be none giving the outward appearance of a compliant publisher to any subscribers. I've marked it internal-only and clearly documented this gotcha/limitation.

Reply:
Also I did run the TCL tests and fixed all the issues that were flagged that were not working as intended.